File: admin/src/theme/themeController.test.ts
import { beforeEach, expect, test } from 'vitest'
import { ThemeController } from './themeController.ts'
import {
  LocalStorageThemeStore,
  themeStorageKey,
  type ThemePreferenceStore,
} from './themePreference.ts'

class MemoryThemePreferenceStore implements ThemePreferenceStore {
  value: string | null

  constructor(value: string | null = null) {
    this.value = value
  }

  read(): string | null {
    return this.value
  }

  write(value: string): void {
    this.value = value
  }
}

beforeEach(() => {
  window.localStorage.clear()
})

test('unknown or empty preference follows the system', () => {
  const store = new MemoryThemePreferenceStore('nope')
  const controller = new ThemeController(store)

  controller.load()

  expect(controller.mode).toBe('system')
})

test('loads a persisted light or dark choice', () => {
  const store = new MemoryThemePreferenceStore('dark')
  const controller = new ThemeController(store)

  controller.load()
  expect(controller.mode).toBe('dark')

  store.value = 'light'
  controller.load()
  expect(controller.mode).toBe('light')
})

test('toggle persists the opposite of the resolved brightness', () => {
  const store = new MemoryThemePreferenceStore()
  const controller = new ThemeController(store)

  controller.toggle('light')
  expect(controller.mode).toBe('dark')
  expect(store.value).toBe('dark')

  controller.toggle('dark')
  expect(controller.mode).toBe('light')
  expect(store.value).toBe('light')
})

test('keeps the toggled choice when storage throws', () => {
  const store: ThemePreferenceStore = {
    read: () => null,
    write: () => {
      throw new Error('storage blocked')
    },
  }
  const controller = new ThemeController(store)

  controller.toggle('light')

  expect(controller.mode).toBe('dark')
})

test('persists the choice in local storage under theme_mode', () => {
  const controller = new ThemeController(new LocalStorageThemeStore())
  controller.load()
  expect(controller.mode).toBe('system')

  controller.toggle('light')
  expect(window.localStorage.getItem(themeStorageKey)).toBe('dark')

  const reloaded = new ThemeController(new LocalStorageThemeStore())
  reloaded.load()
  expect(reloaded.mode).toBe('dark')
})

File: app/test/theme_controller_test.dart
import 'package:eldafttar/src/theme/theme_controller.dart';
import 'package:eldafttar/src/theme/theme_preference_store.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

class MemoryThemePreferenceStore implements ThemePreferenceStore {
  String? value;

  @override
  Future<String?> read() async => value;

  @override
  Future<void> write(String value) async {
    this.value = value;
  }
}

void main() {
  test('unknown or empty preference follows the system', () async {
    final store = MemoryThemePreferenceStore()..value = 'nope';
    final controller = ThemeController(store: store);

    await controller.load();

    expect(controller.mode, ThemeMode.system);
  });

  test('loads a persisted light or dark choice', () async {
    final store = MemoryThemePreferenceStore()..value = 'dark';
    final controller = ThemeController(store: store);

    await controller.load();
    expect(controller.mode, ThemeMode.dark);

    store.value = 'light';
    await controller.load();
    expect(controller.mode, ThemeMode.light);
  });

  test('toggle persists the opposite of the resolved brightness', () async {
    final store = MemoryThemePreferenceStore();
    final controller = ThemeController(store: store);

    await controller.toggle(Brightness.light);
    expect(controller.mode, ThemeMode.dark);
    expect(store.value, 'dark');

    await controller.toggle(Brightness.dark);
    expect(controller.mode, ThemeMode.light);
    expect(store.value, 'light');
  });
}
